precompile/fee_config_manager.go

```diff
@@ -0,0 +1,268 @@
+func PackFeeConfig(feeConfig commontype.FeeConfig) ([]byte, error) {
+	//  input(feeConfig)
+	return packHelper(feeConfig, false)
+}
+
+// PackSetFeeConfig packs [feeConfig] with the selector into the appropriate arguments for setting fee config operations.
+func PackSetFeeConfig(feeConfig commontype.FeeConfig) ([]byte, error) {
+	// function selector (4 bytes) + input(feeConfig)
+	return packHelper(feeConfig, true)
+}
+
+func packHelper(feeConfig commontype.FeeConfig, useSelector bool) ([]byte, error) {
+	fullLen := feeConfigInputLen
+	packed := [][]byte{
+		feeConfig.GasLimit.FillBytes(make([]byte, 32)),
+		new(big.Int).SetUint64(feeConfig.TargetBlockRate).FillBytes(make([]byte, 32)),
+		feeConfig.MinBaseFee.FillBytes(make([]byte, 32)),
+		feeConfig.TargetGas.FillBytes(make([]byte, 32)),
+		feeConfig.BaseFeeChangeDenominator.FillBytes(make([]byte, 32)),
+		feeConfig.MinBlockGasCost.FillBytes(make([]byte, 32)),
+		feeConfig.MaxBlockGasCost.FillBytes(make([]byte, 32)),
+		feeConfig.BlockGasCostStep.FillBytes(make([]byte, 32)),
+	}
+	if useSelector {
+		packed = append([][]byte{setFeeConfigSignature}, packed...)
+		return packOrderedHashesWithSelector(packed, fullLen+selectorLen)
+	}
+
+	return packOrderedHashes(packed, fullLen)
+}
+
+// UnpackFeeConfigInput attempts to unpack [input] into the arguments to the fee config precompile
+// assumes that [input] does not include selector (omits first 4 bytes in PackSetFeeConfigInput)
+func UnpackFeeConfigInput(input []byte) (commontype.FeeConfig, error) {
+	if len(input) != feeConfigInputLen {
+		return commontype.FeeConfig{}, fmt.Errorf("invalid input length for fee config input: %d", len(input))
+	}
+	return commontype.FeeConfig{
+		GasLimit:                 new(big.Int).SetBytes(returnPackedElement(input, 0)),
+		TargetBlockRate:          new(big.Int).SetBytes(returnPackedElement(input, 1)).Uint64(),
+		MinBaseFee:               new(big.Int).SetBytes(returnPackedElement(input, 2)),
+		TargetGas:                new(big.Int).SetBytes(returnPackedElement(input, 3)),
+		BaseFeeChangeDenominator: new(big.Int).SetBytes(returnPackedElement(input, 4)),
+		MinBlockGasCost:          new(big.Int).SetBytes(returnPackedElement(input, 5)),
+		MaxBlockGasCost:          new(big.Int).SetBytes(returnPackedElement(input, 6)),
+		BlockGasCostStep:         new(big.Int).SetBytes(returnPackedElement(input, 7)),
+	}, nil
+}
+
+// GetStoredFeeConfig returns fee config from contract storage in given state
+func GetStoredFeeConfig(stateDB StateDB) (commontype.FeeConfig, bool) {
+	if !stateDB.Exist(FeeConfigManagerAddress) {
+		return commontype.FeeConfig{}, false
+	}
+	feeConfig := commontype.FeeConfig{}
+	isAllZeros := true
+	for i := minFeeConfigFieldKey; i <= maxFeeConfigFieldKey; i++ {
+		val := stateDB.GetState(FeeConfigManagerAddress, common.Hash{byte(i)})
+		// even if one of these vals are non-zero means that this state exist
+		if isAllZeros && val != (common.Hash{0}) {
```

Comment:
nit: isn't `common.Hash{}` already all zeros?

Reply:
it's specifically checking all fields to ensure that all slots contain zeros. This can be used as a `reset`, i.e `setFeeConfig(0,0,0,0,0,0,0,0) == reset()` which reverts back to the genesis config.

Reply:
I think Patrick's original comment meant to say that we can use `common.Hash{}` instead of `common.Hash{0}` here